Ability to use regex when matching on project and branch names

djarosz opened the pull request:
If project or branch name starts with '^' it is treated as regex
otherwise exact match is performed.

Rationale:
- if project names follow some naming convetion
  (e.g. backend/app1, backend/app2, mobile/android/app1,
  mobile/ios/app1) it is easier to configure slack notification for
  "group" of projects like backend, mobile/ios, mobile/android instead
  of manually adding each new Gerrit project to configuration file
  when it is created
- similarly if branches follow naming convention (e.g. release/x.y.z)
  it is easier to match on such branches in generic way. This can be
  also useful e.g. when Gerrit is configured to support per developer
  sandbox branches which usually should be ignored in notifications.

djarosz committed:
Ability to use regex when matching in project and branch names

If project or branch name starts with '^' it is treated as regex
otherwise exact match is performed.

Rationale:
- if project names follow some naming convetion
  (e.g. backend/app1, backend/app2, mobile/android/app1,
  mobile/ios/app1) it is easier to configure slack notification for
  "group" of projects like backend, mobile/ios, mobile/android instead
  of manually adding each new Gerrit project to configuration file
  when it is created
- similarly if branches follow naming convention (e.g. release/x.y.z)
  it is easier to match on such branches in generic way. This can be
  also useful e.g. when Gerrit is configured to support per developer
  sandbox branches which usually should be ignored in notifications.

MichaelSims commented:
Hey, thanks for the pull request.

I can definitely see the rationale for this and think it's a good idea. A couple of issues though. First, the discrepancy between how the project and branch are handled vs. the subject never really occurred to me. With this feature making regex an option for the first two, it sort of highlights that discrepancy and putting on my user hat it makes me wonder why they are handled differently. Also, using "^" as a signifier for "this is a regex" is slightly problematic in that you may want a regex that doesn't start with the "^" anchor. That's not a big deal (it can always be worked around with "^.*" as a prefix) but again it's a discrepancy from the way the subject is handled.

I think I'd prefer if we just enforce that project, branch, and subject are all handled the exact same way. IOW, they are all treated as regexes always (with the special case of "*" being different). What do you think?

djarosz commented:
Sounds like good idea as it will simplify matching and make it more consistent.

One thing  to consider thou as this change will affect users already using gerrit-slack-bot. E.g. where project name was e.g. 'my-project' (matcher value defined in configuration file) after this change this matcher will also match project my-project-new, sandbox/my-project. 

This is probably no big deal since this is relatively young project (I think :)) and secondly this kind of name "clashes" are rather not common.

Do You want me to fix this PR or will You do I yourself?

MichaelSims commented:
Good point, it is a BC break, but according to Github I've had 6 unique visitors to this repo and not counting us, that leaves 4 people. I'll include a prominent warning in the release description. :-)

If you are able to fix the PR that would be great. If not, I can get to it in a few days most likely.

djarosz committed:
Matcher for project/branch/subject are always treated as regexes

Project, branch, and subject are all handled the exact same way.
They are all treated as regexes always (with the special case of "*"
being different).

MichaelSims commented on app/src/main/java/sims/michael/gerritslackbot/model/ChangeMatcher.kt:
```diff
@@ -29,10 +29,8 @@ data class ChangeMatcher(
-        else -> pattern.toLowerCase() in listOf("*", this.toLowerCase())
-    }
+    private fun String?.safeMatchesWildcardOrRegex(pattern: String) =
+            pattern == "*" || this?.safeMatches(pattern) ?: true
```
The last boolean in this expression should be `false`. As it is, if a change subject is missing for some reason (null) then any subject pattern provided will match it. A missing subject should be matched only by `"*"`

djarosz replied:
Good point.

MichaelSims reviewed (changes requested):
Looks good, thanks! Just one tweak to make.

djarosz committed:
Regex match on empty event subject should be negative
